// azure-pipelines-tasks-codecoverage-tools/jacoco/jacoco.maven.ccenabler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JacocoMavenCodeCoverageEnabler = void 0;
const util = require("../utilities");
const tl = require("azure-pipelines-task-lib/task");
const ccc = require("../codecoverageconstants");
const cc = require("../codecoverageenabler");
const Q = require("q");
const path = require("path");
tl.setResourcePath(path.join(path.dirname(__dirname), 'module.json'), true);
class JacocoMavenCodeCoverageEnabler extends cc.JacocoCodeCoverageEnabler {
    // -----------------------------------------------------
    // Enable code coverage for Jacoco Maven Builds
    // - enableCodeCoverage: CodeCoverageProperties  - ccProps
    // -----------------------------------------------------
    enableCodeCoverage(ccProps) {
        let _this = this;
        tl.debug("Input parameters: " + JSON.stringify(ccProps));
        _this.buildFile = ccProps["buildfile"];
        _this.reportDir = ccProps["reportdirectory"];
        _this.sourceDirs = ccProps["sourcedirectories"];
        _this.classDirs = ccProps["classfilesdirectories"];
        _this.reportBuildFile = ccProps["reportbuildfile"];
        let classFilter = ccProps["classfilter"];
        let filter = _this.extractFilters(classFilter);
        _this.excludeFilter = _this.applyFilterPattern(filter.excludeFilter);
        _this.includeFilter = _this.applyFilterPattern(filter.includeFilter);
        return util.readXmlFileAsJson(_this.buildFile)
            .then(function (resp) {
            return _this.addCodeCoverageData(resp);
        });
    }
    applyFilterPattern(filter) {
        let ccfilter = [];
        if (!util.isNullOrWhitespace(filter)) {
            util.trimToEmptyString(filter).replace(/\./g, "/").split(":").forEach(exFilter => {
                if (exFilter) {
                    ccfilter.push(exFilter.endsWith("*") ? ("**/" + exFilter + "/**") : ("**/" + exFilter + ".class"));
                }
            });
        }
        tl.debug("Applying the filter pattern: " + filter + " op: " + ccfilter);
        return ccfilter;
    }
    addCodeCoverageData(pomJson) {
        let _this = this;
        if (!pomJson.project) {
            Q.reject(tl.loc("InvalidBuildFile"));
        }
        let isMultiModule = false;
        const originalPom = JSON.parse(JSON.stringify(pomJson));
        if (pomJson.project.modules) {
            tl.debug("Multimodule project detected");
            isMultiModule = true;
        }
        let promises = [_this.addCodeCoveragePluginData(pomJson)];
        if (isMultiModule) {
            pomJson.project.modules[0].module.push(path.basename(_this.reportDir));
            promises.push(_this.createMultiModuleReport(originalPom));
        }
        return Q.all(promises)
            .then(() => {
            const jacocoDir = isMultiModule ?
                "jacoco-aggregate" :
                "jacoco";
            const dirName = isMultiModule ?
                _this.reportDir :
                path.dirname(_this.reportDir);
            return path.join(dirName, "target", "site", jacocoDir);
        });
    }
    addCodeCoverageNodes(buildJsonContent) {
        let _this = this;
        let buildNode = _this.getBuildDataNode(buildJsonContent);
        let pluginsNode = _this.getPluginDataNode(buildNode);
        let ccContent = ccc.jacocoMavenPluginEnable(_this.includeFilter, _this.excludeFilter);
        util.addPropToJson(pluginsNode, "plugin", ccContent);
        return Q.resolve(buildJsonContent);
    }
    getBuildDataNode(buildJsonContent) {
        let buildNode = null;
        if (!buildJsonContent.project.build || typeof buildJsonContent.project.build === "string") {
            buildNode = {};
            buildJsonContent.project.build = buildNode;
        }
        else if (buildJsonContent.project.build instanceof Array) {
            if (typeof buildJsonContent.project.build[0] === "string") {
                buildNode = {};
                buildJsonContent.project.build[0] = buildNode;
            }
            else {
                buildNode = buildJsonContent.project.build[0];
            }
        }
        return buildNode;
    }
    getPluginDataNode(buildNode) {
        let pluginsNode = {};
        /* Always look for plugins node first */
        if (buildNode.plugins) {
            if (typeof buildNode.plugins === "string") {
                buildNode.plugins = {};
            }
            if (buildNode.plugins instanceof Array) {
                if (typeof buildNode.plugins[0] === "string") {
                    pluginsNode = {};
                    buildNode.plugins[0] = pluginsNode;
                }
                else {
                    pluginsNode = buildNode.plugins[0];
                }
            }
            else {
                pluginsNode = buildNode.plugins;
            }
        }
        else {
            buildNode.plugins = {};
            pluginsNode = buildNode.plugins;
        }
        return pluginsNode;
    }
    getParentPomData(pomJson) {
        var _a;
        const groupId = pomJson.project.groupId[0];
        const artifactId = pomJson.project.artifactId[0];
        const version = pomJson.project.version[0];
        const jsonModulesProperty = (_a = pomJson.project) === null || _a === void 0 ? void 0 : _a.modules;
        const modules = jsonModulesProperty instanceof Array ?
            jsonModulesProperty[0].module :
            [];
        return {
            groupId,
            artifactId,
            version,
            modules
        };
    }
    formatParentData(data) {
        return `
                <parent>
                    <groupId>${data.groupId}</groupId>
                    <artifactId>${data.artifactId}</artifactId>
                    <version>${data.version}</version>
                </parent>
                `;
    }
    formatModulesData(modules) {
        const dependencies = modules.reduce((acc, current) => {
            return `
                    ${acc}
                    <dependency>
                        <groupId>${current.groupId}</groupId>
                        <artifactId>${current.artifactId}</artifactId>
                        <version>${current.version}</version>
                        <type>${current.packaging}</type>
                    </dependency>
                    `;
        }, '');
        return `<dependencies>
                    ${dependencies}
                </dependencies>`;
    }
    getModulesData(data) {
        const _this = this;
        return data.modules.map((moduleName) => {
            const modulePom = path.join(path.dirname(_this.reportDir), moduleName, 'pom.xml');
            return util.readXmlFileAsJson(modulePom)
                .then((content) => {
                var _a, _b, _c;
                const jsonGroup = (_a = content.project) === null || _a === void 0 ? void 0 : _a.groupId;
                const jsonVersion = (_b = content.project) === null || _b === void 0 ? void 0 : _b.version;
                const jsonPackaging = (_c = content.project) === null || _c === void 0 ? void 0 : _c.packaging;
                return {
                    artifactId: content.project.artifactId[0],
                    groupId: (jsonGroup && jsonGroup[0]) || data.groupId,
                    version: (jsonVersion && jsonVersion[0]) || data.version,
                    packaging: (jsonPackaging && jsonPackaging[0]) || "jar"
                };
            });
        });
    }
    createMultiModuleReport(pomJson) {
        let _this = this;
        let srcDirs = _this.sourceDirs;
        let classDirs = _this.classDirs;
        let includeFilter = _this.includeFilter.join(",");
        let excludeFilter = _this.excludeFilter.join(",");
        const reportArtifactId = path.basename(_this.reportDir);
        const parentData = _this.getParentPomData(pomJson);
        if (util.isNullOrWhitespace(srcDirs)) {
            srcDirs = ".";
        }
        if (util.isNullOrWhitespace(classDirs)) {
            classDirs = ".";
        }
        return Q.all(_this.getModulesData(parentData))
            .then((modules) => {
            return util.writeFile(_this.reportBuildFile, ccc.jacocoMavenMultiModuleReport(reportArtifactId, srcDirs, classDirs, includeFilter, excludeFilter, parentData.groupId, _this.formatParentData(parentData), _this.formatModulesData(modules)));
        });
    }
    addCodeCoveragePluginData(pomJson) {
        let _this = this;
        return _this.addCodeCoverageNodes(pomJson)
            .then(function (content) {
            return util.writeJsonAsXmlFile(_this.buildFile, content);
        });
    }
}
exports.JacocoMavenCodeCoverageEnabler = JacocoMavenCodeCoverageEnabler;
