// azure-pipelines-tasks-codecoverage-tools/codecoverageconstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.coberturaGradleBuildScript = exports.coberturaAntClasspathRef = exports.coberturaAntProperties = exports.coberturaAntInstrumentedClasses = exports.coberturaAntCoverageEnable = exports.coberturaAntReport = exports.jacocoAntCoverageEnable = exports.jacocoAntReport = exports.coberturaMavenReport = exports.coberturaMavenEnable = exports.jacocoMavenMultiModuleReport = exports.jacocoMavenPluginEnable = exports.coberturaGradleMultiModuleEnable = exports.coberturaGradleSingleModuleEnable = exports.jacocoGradleSingleModuleEnableV2 = exports.jacocoGradleSingleModuleEnable = exports.jacocoGradleMultiModuleEnableV2 = exports.jacocoGradleMultiModuleEnable = void 0;
const path = require("path");
const util = require("./utilities");
const os = require("os");
/**
 * Returns assignment expression for specified property depending on Gradle version
 * Starting from Gradle 5 there is a breaking change with assignment - '=' character no longer works, 'setFrom' method is required instead
 *
 * For more info - please check link
 *
 * @param {string} propretyToAssign prorety to assign value to
 * @param {boolean} gradle5xAndHigher set to true if Gradle version is higher that 5; should be 'false' otherwise
 * @returns {string} formatted assignment expression (for example, 'property=' or 'property.setFrom')
 */
function getFormattedFileCollectionAssignGradle(propretyToAssign, gradle5xAndHigher) {
    if (gradle5xAndHigher) {
        return `${propretyToAssign}.setFrom`;
    }
    else {
        return `${propretyToAssign} =`;
    }
}
// Enable Jacoco Code Coverage for Gradle builds using this props
function jacocoGradleMultiModuleEnable(excludeFilter, includeFilter, classFileDirectory, reportDir, gradle5xOrHigher) {
    return `
allprojects {
    repositories {
        mavenCentral()
    }

    apply plugin: 'jacoco'
}

def jacocoExcludes = [${excludeFilter}]
def jacocoIncludes = [${includeFilter}]

subprojects {
    jacocoTestReport {
        doFirst {
            ${getFormattedFileCollectionAssignGradle('classDirectories', gradle5xOrHigher)} fileTree(dir: "${classFileDirectory}").exclude(jacocoExcludes).include(jacocoIncludes)
        }

        reports {
            html.enabled = true
            html.destination file("\${buildDir}/jacocoHtml")
            xml.enabled = true
            xml.destination file("\${buildDir}/summary.xml")
        }
    }
    test {
        jacoco {
            destinationFile = file("${reportDir}/jacoco.exec")
        }
    }
}

task jacocoRootReport(type: org.gradle.testing.jacoco.tasks.JacocoReport) {
    dependsOn = subprojects.test
    ${getFormattedFileCollectionAssignGradle('executionData', gradle5xOrHigher)} files(subprojects.jacocoTestReport.executionData)
    ${getFormattedFileCollectionAssignGradle('sourceDirectories', gradle5xOrHigher)} files(subprojects.sourceSets.main.allSource.srcDirs)
    ${getFormattedFileCollectionAssignGradle('classDirectories', gradle5xOrHigher)} files()

    doFirst {
        subprojects.each {
            if (new File("\${it.sourceSets.main.output.classesDirs}").exists()) {
                logger.info("Class directory exists in sub project: \${it.name}")
                logger.info("Adding class files \${it.sourceSets.main.output.classesDirs}")
                classDirectories += fileTree(dir: "\${it.sourceSets.main.output.classesDirs}", includes: jacocoIncludes, excludes: jacocoExcludes)
            } else {
                logger.error("Class directory does not exist in sub project: \${it.name}")
            }
        }
    }

    reports {
        html.enabled = true
        xml.enabled = true
        xml.destination file("${reportDir}/summary.xml")
        html.destination file("${reportDir}/")
    }
}`;
}
exports.jacocoGradleMultiModuleEnable = jacocoGradleMultiModuleEnable;
// Enable Jacoco Code Coverage for multi-module Gradle projects (for Gradle version 6 and higher). 
function jacocoGradleMultiModuleEnableV2(excludeFilter, includeFilter, classFileDirectory, reportDir) {
    return `
allprojects {
  repositories {
      mavenCentral()
  }

  apply plugin: 'jacoco'
}

def jacocoExcludes = [${excludeFilter}]
def jacocoIncludes = [${includeFilter}]

subprojects {
  jacocoTestReport {
      afterEvaluate {
          classDirectories.setFrom files(
                  fileTree(dir: "${classFileDirectory}", exclude: jacocoExcludes, include: jacocoIncludes)
          )
      }
      reports {
          html.required = true
          html.destination file("\${buildDir}/jacocoHtml")
          xml.required = true
          xml.destination file("\${buildDir}/summary.xml")
      }
  }
  test {
      jacoco {
          destinationFile = file("${reportDir}/jacoco.exec")
      }
  }
}
task jacocoRootReport(type: JacocoReport) {
  dependsOn = subprojects.test
  executionData.setFrom files(subprojects.jacocoTestReport.executionData)
  sourceDirectories.setFrom files(subprojects.sourceSets.main.allSource.srcDirs)
  classDirectories.setFrom files(subprojects.sourceSets.main.output)
  afterEvaluate {
      classDirectories.setFrom(
              files(classDirectories.files.collect {
                  fileTree(dir: it, exclude: jacocoExcludes)
              })
      )
  }
  reports {
      html.required = true
      xml.required = true
      xml.destination file("${reportDir}/summary.xml")
      html.destination file("${reportDir}/")
  }
}`;
}
exports.jacocoGradleMultiModuleEnableV2 = jacocoGradleMultiModuleEnableV2;
function jacocoGradleSingleModuleEnable(excludeFilter, includeFilter, classFileDirectory, reportDir, gradle5xOrHigher) {
    return `
allprojects {
    repositories {
        mavenCentral()
    }

    apply plugin: 'jacoco'
}

def jacocoExcludes = [${excludeFilter}]
def jacocoIncludes = [${includeFilter}]

jacocoTestReport {
    doFirst {
        ${getFormattedFileCollectionAssignGradle('classDirectories', gradle5xOrHigher)} fileTree(dir: "${classFileDirectory}").exclude(jacocoExcludes).include(jacocoIncludes)
    }

    reports {
        html.enabled = true
        xml.enabled = true
        xml.destination file("${reportDir}/summary.xml")
        html.destination file("${reportDir}")
    }
}

test {
    finalizedBy jacocoTestReport
    jacoco {
        destinationFile = file("${reportDir}/jacoco.exec")
    }
}`;
}
exports.jacocoGradleSingleModuleEnable = jacocoGradleSingleModuleEnable;
// Enable Jacoco Code Coverage for single-module Gradle projects (for Gradle version 6 and higher). 
function jacocoGradleSingleModuleEnableV2(excludeFilter, includeFilter, classFileDirectory, reportDir, gradle5xOrHigher) {
    return `
allprojects {
    repositories {
        mavenCentral()
    }

    apply plugin: 'jacoco'
}

def jacocoExcludes = [${excludeFilter}]
def jacocoIncludes = [${includeFilter}]

jacocoTestReport {
    afterEvaluate {
        classDirectories.setFrom files(
            fileTree(dir: "${classFileDirectory}", exclude: jacocoExcludes, include: jacocoIncludes)
        )
    }

    reports {
        html.required = true
        xml.required = true
        xml.destination file("${reportDir}/summary.xml")
        html.destination file("${reportDir}")
    }
}

test {
    finalizedBy jacocoTestReport
    jacoco {
        destinationFile = file("${reportDir}/jacoco.exec")
    }
}`;
}
exports.jacocoGradleSingleModuleEnableV2 = jacocoGradleSingleModuleEnableV2;
// Enable Cobertura Code Coverage for Gradle builds using this props
function coberturaGradleSingleModuleEnable(excludeFilter, includeFilter, classDir, sourceDir, reportDir) {
    if (!classDir) {
        classDir = "${project.sourceSets.main.output.classesDirs}";
    }
    if (!sourceDir) {
        sourceDir = "project.sourceSets.main.java.srcDirs";
    }
    return `
allprojects {
    repositories {
        mavenCentral()
    }
    apply plugin: 'net.saliman.cobertura'

    dependencies {
        testCompile 'org.slf4j:slf4j-api:1.7.12'
    }

    cobertura.coverageIncludes = [${includeFilter}]
    cobertura.coverageExcludes = [${excludeFilter}]
}

cobertura {
    coverageDirs = ["${classDir}"]
    coverageSourceDirs = ${sourceDir}
    coverageReportDir = new File('${reportDir}')
    coverageFormats = ['xml', 'html']
}`;
}
exports.coberturaGradleSingleModuleEnable = coberturaGradleSingleModuleEnable;
function coberturaGradleMultiModuleEnable(excludeFilter, includeFilter, classDir, sourceDir, reportDir) {
    let data = `
allprojects {
    repositories {
        mavenCentral()
    }
    apply plugin: 'net.saliman.cobertura'

    dependencies {
        testCompile 'org.slf4j:slf4j-api:1.7.12'
    }

    cobertura.coverageIncludes = [${includeFilter}]
    cobertura.coverageExcludes = [${excludeFilter}]
}

test {
    dependsOn = subprojects.test
}

cobertura {
    coverageSourceDirs = []`;
    if (classDir) {
        data += `
        coverageDirs = ["${classDir}"]`;
    }
    else {
        data += `
    rootProject.subprojects.each {
        coverageDirs << file("\${it.sourceSets.main.output.classesDirs}")
    }`;
    }
    if (sourceDir) {
        data += `
    coverageDirs = ["${sourceDir}"]`;
    }
    else {
        data += `
    rootProject.subprojects.each {
        coverageSourceDirs += it.sourceSets.main.java.srcDirs
    }`;
    }
    data += `
    coverageFormats = [ 'xml', 'html' ]
    coverageMergeDatafiles = subprojects.collect { new File(it.projectDir, '/build/cobertura/cobertura.ser') }
    coverageReportDir = new File('${reportDir}')
}`;
    return data;
}
exports.coberturaGradleMultiModuleEnable = coberturaGradleMultiModuleEnable;
;
// Enable Jacoco Code Coverage for Maven builds using this props
function jacocoMavenPluginEnable(includeFilter, excludeFilter) {
    let plugin = {
        "groupId": "org.jacoco",
        "artifactId": "jacoco-maven-plugin",
        "version": "0.8.11",
        "configuration": {
            "includes": [{
                    "include": includeFilter,
                }],
            "excludes": [{
                    "exclude": excludeFilter
                }]
        },
        "executions": {
            "execution": [
                {
                    "configuration": {
                        "includes": [{
                                "include": "**/*",
                            }]
                    },
                    "id": "default-prepare-agent-vsts",
                    "goals": { "goal": "prepare-agent" }
                },
                {
                    "id": "default-report-vsts",
                    "goals": { "goal": "report" },
                    "phase": "test"
                }
            ]
        }
    };
    return plugin;
}
exports.jacocoMavenPluginEnable = jacocoMavenPluginEnable;
;
function jacocoMavenMultiModuleReport(reportArtifactId, srcData, classData, includeFilter, excludeFilter, groupId, parentData, modules) {
    let classNode = "";
    classData.split(",").forEach(c => {
        classNode += `<fileset dir="${c}"`;
        if (includeFilter) {
            classNode += ` includes="${includeFilter}"`;
        }
        if (excludeFilter) {
            classNode += ` excludes="${excludeFilter}"`;
        }
        classNode += ` />` + os.EOL;
    });
    let srcNode = "";
    if (util.isNullOrWhitespace(srcData)) {
        srcNode = `<fileset dir="." />`;
    }
    else {
        srcData.split(",").forEach(c => {
            srcNode += `<fileset dir="${c}" />` + os.EOL;
        });
    }
    let report = `<?xml version="1.0" encoding="UTF-8"?>
        <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
            <modelVersion>4.0.0</modelVersion>
            <groupId>${groupId}</groupId>
            <artifactId>${reportArtifactId}</artifactId>
            <version>1.0-SNAPSHOT</version>
            <packaging>pom</packaging>
            ${modules}
            ${parentData}
            <build>
                <plugins>
                    <plugin>
                        <groupId>org.jacoco</groupId>
                        <artifactId>jacoco-maven-plugin</artifactId>
                        <version>0.8.11</version>
                        <executions>
                            <execution>
                                <id>jacoco-report-aggregate</id>
                                <phase>verify</phase>
                                <goals>
                                    <goal>report-aggregate</goal>
                                </goals>
                            </execution>
                        </executions>
                    </plugin>
                </plugins>
            </build>
        </project>`;
    return report;
}
exports.jacocoMavenMultiModuleReport = jacocoMavenMultiModuleReport;
;
// Enable Cobertura Code Coverage for Maven builds using this props
function coberturaMavenEnable(includeFilter, excludeFilter, aggregate) {
    let includeTag = "";
    let excludeTag = "";
    if (!util.isNullOrWhitespace(excludeFilter)) {
        excludeFilter.split(",").forEach(ex => {
            excludeTag += `<exclude>${ex}</exclude>` + os.EOL;
        });
    }
    if (!util.isNullOrWhitespace(includeFilter)) {
        includeFilter.split(",").forEach(ex => {
            includeTag += `<include>${ex}</include>` + os.EOL;
        });
    }
    let ccProperty = `
    <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>cobertura-maven-plugin</artifactId>
        <version>2.7</version>
        <configuration>
          <formats>
            <format>xml</format>
            <format>html</format>
          </formats>
          <instrumentation>
            <includes>${includeTag}</includes>
            <excludes>${excludeTag}</excludes>
          </instrumentation>
          <aggregate>${aggregate}</aggregate>
        </configuration>
        <executions>
          <execution>
            <id>package-9af52907-6506-4b87-b16a-9883edee41bc</id>
            <goals>
              <goal>cobertura</goal>
            </goals>
            <phase>package</phase>
          </execution>
        </executions>
    </plugin>
  `;
    return util.convertXmlStringToJson(ccProperty);
}
exports.coberturaMavenEnable = coberturaMavenEnable;
;
function coberturaMavenReport() {
    let ccProperty = `
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>cobertura-maven-plugin</artifactId>
        <version>2.7</version>
        <configuration>
          <formats>
            <format>xml</format>
            <format>html</format>
          </formats>
        </configuration>
      </plugin>
  `;
    return util.convertXmlStringToJson(ccProperty);
}
exports.coberturaMavenReport = coberturaMavenReport;
function jacocoAntReport(reportDir, classData, sourceData) {
    return `<?xml version="1.0"?>
<project name="JacocoReport">
    <target name="CodeCoverage_9064e1d0">
        <jacoco:report xmlns:jacoco="antlib:org.jacoco.ant">
            <executiondata>
                <file file="${path.join(reportDir, "jacoco.exec")}"/>
            </executiondata>
            <structure name="Jacoco report">
                <classfiles>${classData}</classfiles>
                <sourcefiles>${sourceData}</sourcefiles>
            </structure>
            <html destdir="${reportDir}" />
            <csv destfile="${reportDir + path.sep}summary.csv" />
            <xml destfile="${reportDir + path.sep}summary.xml" />
        </jacoco:report>
    </target>
</project>
    `;
}
exports.jacocoAntReport = jacocoAntReport;
function jacocoAntCoverageEnable(reportDir) {
    let file = path.join(reportDir, "jacoco.exec");
    return {
        $: {
            "destfile": file,
            "xmlns:jacoco": "antlib:org.jacoco.ant"
        }
    };
}
exports.jacocoAntCoverageEnable = jacocoAntCoverageEnable;
function coberturaAntReport(srcDir, reportDir) {
    return `<?xml version="1.0"?>
<project name="CoberturaReport">
  <property environment="env" />
  <path id="cobertura-classpath" description="classpath for instrumenting classes">
    <fileset dir="\${env.COBERTURA_HOME}">
      <include name="cobertura*.jar" />
      <include name="**/lib/**/*.jar" />
    </fileset>
  </path>
  <taskdef classpathref="cobertura-classpath" resource="tasks.properties" />
  <target name="CodeCoverage_9064e1d0">
    <cobertura-report format="html" destdir="${reportDir}" datafile="${reportDir + path.sep}cobertura.ser" srcdir="${srcDir}" />
    <cobertura-report format="xml" destdir="${reportDir}" datafile="${reportDir + path.sep}cobertura.ser" srcdir="${srcDir}" />
  </target>
</project>
    `;
}
exports.coberturaAntReport = coberturaAntReport;
function coberturaAntCoverageEnable() {
    return `
<property environment="env" />
<path id="cobertura-classpath-vsts" description="classpath for instrumenting classes">
    <fileset dir="\${env.COBERTURA_HOME}">
        <include name="cobertura*.jar" />
        <include name="**/lib/**/*.jar" />
    </fileset>
</path>
<taskdef classpathref="cobertura-classpath-vsts" resource="tasks.properties" />
`;
}
exports.coberturaAntCoverageEnable = coberturaAntCoverageEnable;
function coberturaAntInstrumentedClasses(baseDir, reportDir, classData) {
    return `
<cobertura-instrument todir="${path.join(baseDir, "InstrumentedClasses")}" datafile="${path.join(reportDir, "cobertura.ser")}">
    ${classData}
</cobertura-instrument>
  `;
}
exports.coberturaAntInstrumentedClasses = coberturaAntInstrumentedClasses;
function coberturaAntProperties(reportDir, baseDir) {
    return `
        <sysproperty key="net.sourceforge.cobertura.datafile" file="${path.join(reportDir, "cobertura.ser")}" />
        <classpath location="${path.join(baseDir, "InstrumentedClasses")}" />
`;
}
exports.coberturaAntProperties = coberturaAntProperties;
function coberturaAntClasspathRef() {
    return `
        <classpath refid="cobertura-classpath-vsts" />
`;
}
exports.coberturaAntClasspathRef = coberturaAntClasspathRef;
// Gradle Coberutra plugin
exports.coberturaGradleBuildScript = `
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'net.saliman:gradle-cobertura-plugin:2.2.7'
    }
}
`;
